src/pymor/reductors/neural_network.py

```diff
@@ -124,19 +125,17 @@ def reduce(self, hidden_layers='[(N+P)*3, (N+P)*3]', activation_function=torch.t
             # input and output size of the neural network are prescribed by the dimension of the parameter space
             # and the reduced basis size
             assert isinstance(hidden_layers, list)
-            layers = [len(self.fom.parameters),] + hidden_layers + [len(self.reduced_basis),]
+            layers = self._compute_layers_sizes(hidden_layers)
 
             # compute validation data
             if not hasattr(self, 'validation_data'):
                 with self.logger.block('Computing validation snapshots ...'):
 
                     if self.validation_set:
-                        self.validation_data = []
+                        U = self.fom.solution_space.empty()
                         for mu in self.validation_set:
-                            mu_tensor = torch.DoubleTensor(mu.to_numpy())
-                            u = self.fom.solve(mu)
-                            u_tensor = torch.DoubleTensor(self.reduced_basis.inner(u)[:,0])
-                            self.validation_data.append((mu_tensor, u_tensor))
+                            U.append(self.fom.solve(mu))
+                        self.validation_data = self._compute_samples(self.validation_set, U, self.reduced_basis)
```

Comment:
I don't like that approach. This forces you to keep all validation snapshots in memory even though you only need their projections. Maybe pass a callable or generator to `_compute_samples` instead of `U`?

Reply:
I have to think about a good solution for this. The same applies for the creation of the training data in `build_basis`, right?

Reply:
The situation there is different: For a standard POD, you need to have all snapshot data in memory. One could go with a HAPOD/incremental SVD, but that might be overkill at the moment. In the case of the validation snapshots, however, the problem can be easily avoided.

Reply:
Okay, I see. So it should be possible to pass either a VectorArray `U`, containing the solutions to the parameter set, or a callable like `self.fom.solve` to `_compute_samples`?

Reply:
Maybe the most elegant way would be to pass an iterable. Then you can directly pass the `VectorArray` in `build_basis` and pass a generator for the validation snapshots.

Reply:
Oh, sorry, won't work.

Reply:
Maybe you should just move the loop out of `_compute_samples` to the caller. I think this might help.

Reply:
I am not completely happy with the new solution in ca35785, maybe you have a better idea @sdrave?